### PR TITLE
Sale Tax Problem

### PR DESCRIPTION
**Pull request by Shubham-Porwall**
Creating a pull request for reviewing of sales-tax code

### CODE REVIEW

File: input.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class Inp
```
Comment (kalpana19sharma) [inline input.rb]: Use proper class and method names.

File: input.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  def inp
+      @arr = []
+      @ch
```
Comment (kalpana19sharma) [inline input.rb]: Use proper variable names

File: input.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+      begin
+          puts "Enter the values"
+          str = gets.chomp
+          if ( str.split(/ /).first.to_f.negative?() || str.split(/ /).last.to_f.negative?() )
```
Comment (kalpana19sharma) [inline input.rb]: I think it's 4 space indentation.

File: input.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+          else
+          @arr.push(str)
+          end
+          puts "Quit = 'q'"
+          ch = gets.chomp
+      end while ch != 'q'
+      @arr
+  end
+end
+
```
Comment (kalpana19sharma) [inline input.rb]: Remove extra empty lines. Just add one new line at the end of the file.

File: new_arr.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  Taxes = obj.ret_tax
+  New_arr = []
+  @@j = 0
+  def disp(arr)
+    for i in (0...arr.length)
+      y = ""
+      x = arr[i].split(/ /)
+      for j in (0...x.length-2)
+        y = y + x[j] + " "
+      end
```
Comment (kalpana19sharma) [inline new_arr.rb]: Use meaning-full variable, classes, modules, method names. Please fix this in the whole code.

File: new_arr.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
+require_relative 'tax_calculate'
+module New
+  obj = Tax.new
+  Taxes = obj.ret_tax
+  New_arr = []
```
Comment (kalpana19sharma) [inline new_arr.rb]: Use correct naming convention.

File: tax_determination.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
+module Determine
```
Comment (kalpana19sharma) [inline tax_determination.rb]: meaning full name

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: input.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
+class Input
+  def read_input
+    @input_store = []
+    @for_quit
```
Comment (kalpana19sharma) [inline input.rb]: Why do we need instance variable here. Local variable should be enough, no?

File: tax_determination.rb [hunk 9]
```diff
@@ -0,0 +1,17 @@
+module Tax_Determination
```
Comment (kalpana19sharma) [inline tax_determination.rb]: https://github.com/github/rubocop-github/blob/master/STYLEGUIDE.md#naming

File: input.rb [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    begin
+      puts "Enter the values"
+      inputs = gets.chomp
```
Comment (kalpana19sharma) [inline input.rb]: why inputs? I mean why plural?

File: input.rb [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+      if ( inputs.split(/ /).first.to_f.negative?() || inputs.split(/ /).last.to_f.negative?() )
```
Comment (kalpana19sharma) [inline input.rb]: 1. We can store `inputs.split(/ /)` in a variable.
2. No need to add empty `()` with method call.
```suggestion
      if ( inputs.split(/ /).first.to_f.negative? || inputs.split(/ /).last.to_f.negative? )
```

File: input.rb [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+        puts "Price and value can't be negative"
+        exit
+        else
+        @input_store.push(inputs)
+      end
```
Comment (kalpana19sharma) [inline input.rb]: Fix indentation.

File: main.rb [hunk 13]
```diff
@@ -0,0 +1,10 @@
+require_relative "input"
+require_relative "sort"
+require_relative "result.rb"
+
+user = Input.new
+arr = user.read_input
```
Comment (kalpana19sharma) [inline main.rb]: Use a meaning full name instead of `arr`.

File: new_arr.rb [hunk 14]
```diff
@@ -0,0 +1,21 @@
+require_relative 'tax_calculate'
+module Create_Array
+  obj = Calculate_Tax.new
+  TAXES = obj.return_all_tax_value
+  NEW_ARR = []
+  @@j = 0
+  def return_result(arr)
+    for i in (0...arr.length)
```
Comment (kalpana19sharma) [inline new_arr.rb]: Since we are processing array elements instead of index so you should use `each` loop on array.

File: new_arr.rb [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+        temp = temp + x[j] + " "
+      end
+      temp = temp + ": " + Create_Array::TAXES[@@j].to_s
+      @@j += 1
+      Create_Array::NEW_ARR << temp
+    end
+    Create_Array::NEW_ARR
+  end
+end
+
```
Comment (kalpana19sharma) [inline new_arr.rb]: Fix empty line.

File: result.rb [hunk 16]
```diff
@@ -0,0 +1,14 @@
+require_relative 'tax_calculate'
+require_relative 'new_arr'
+class Result
+  include Create_Array
+  RESULT = []
+  def display(arr)
+    obj = Calculate_Tax.new
+    arr1 = return_result(arr)
+    Result::RESULT << arr1
```
Comment (kalpana19sharma) [inline result.rb]: I think local variable should be enough here. No need of constant.

File: sort.rb [hunk 17]
```diff
@@ -0,0 +1,25 @@
+require_relative 'tax_determination'
+require_relative 'tax_calculate'
+
+class Logic
+include Tax_Determination
+  def data_sort(arr)
+    obj2 = Calculate_Tax.new
+    for i in (0...arr.length)
```
Comment (kalpana19sharma) [inline sort.rb]: Use `each` loop.

File: sort.rb [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+      @last = arr[i].split(/ /).last.to_f
```
Comment (kalpana19sharma) [inline sort.rb]: Use local variable. We use instance variables when we need to access the variable across the methods.

File: tax_calculate.rb [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+  @@sum_without_tax = 0.0
+  @@sum_with_tax = 0.0
+  @@diff = 0.0
+  def calculate_tax(last, tax)
+    @@sum_without_tax = @@sum_without_tax.to_f + last
+    @value = last + last * tax
+    @@sum_with_tax = @@sum_with_tax + @value 
+    Calculate_Tax::WITH_TAX << @value.round(2)
+  end
+  def return_all_tax_value
```
Comment (kalpana19sharma) [inline tax_calculate.rb]: Add an empty line between method definition.

File: tax_determination.rb [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+  PRODUCT = %w(pills book books chocolate chocolates)
+  def isExempt(arr)
+    @bool = false
```
Comment (kalpana19sharma) [inline tax_determination.rb]: Use local variable.

File: tax_determination.rb [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+      if arr.include?Tax_Determination::PRODUCT[i]
+        @bool = true
+        break;
+      end
+    end
+    if (@bool)
+      return true
+    else
+      return false
+    end
```
Comment (kalpana19sharma) [inline tax_determination.rb]: You can directly return `@bool`.

File: tax_determination.rb [hunk 22]
```diff
@@ -0,0 +1,17 @@
+module Tax_Determination
+  PRODUCT = %w(pills book books chocolate chocolates)
+  def isExempt(arr)
+    @bool = false
+    for i in (0...Tax_Determination::PRODUCT.length)
+      if arr.include?Tax_Determination::PRODUCT[i]
+        @bool = true
+        break;
+      end
+    end
```
Comment (kalpana19sharma) [inline tax_determination.rb]: Try to find a simpler way to do this in ruby.

File: tax_determination.rb [hunk 23]
```diff
@@ -0,0 +1,17 @@
+module Tax_Determination
+  PRODUCT = %w(pills book books chocolate chocolates)
+  def isExempt(arr)
```
Comment (kalpana19sharma) [inline tax_determination.rb]: Methods that return boolean values should end with `?`. And use `snake_case` for method and variable names. Use meaning full parameter names.
```suggestion
  def is_exempt?(products)
```

Comment (kalpana19sharma) [reply tax_determination.rb]: I think this method should handle one product at a time since you are calling this method for each product.

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: input.rb (hunk 10 above)
Comment (Shubham-Porwall) [reply input.rb]: We are taking every string into that local variable and than push into array that's why indicate this as inputs.

File: sort.rb [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+        else
+          obj2.calculate_tax(last, 0.15)
+        end
+      else
+        if is_exempt?(i)
+          obj2.calculate_tax(last, 0.00)  
+        else
+          obj2.calculate_tax(last, 0.10)
+        end
+      end
```
Comment (kalpana19sharma) [inline sort.rb]:
```suggestion
      tax_rate = is_exempt?(i) ? 0.0 : 0.10
      tax_rate += 0.5 if i.include?('imported')
      obj2.calculate_tax(last, tax_rate) 
```
Will this logic work?

File: tax_determination.rb [hunk 25]
```diff
@@ -0,0 +1,13 @@
...
+  def is_exempt?(product)
+    products = %w(pills book books chocolate chocolates)
+    bool = false
+    products.each do |i|
+      if product.include?i
+        bool = true
+        break
+      end
+    end
+    bool
```
Comment (kalpana19sharma) [inline tax_determination.rb]:
```suggestion
    return (product.split & products).present?
```
Will this return the correct value?

File: tax_determination.rb [hunk 26]
```diff
@@ -0,0 +1,13 @@
+module TaxDetermination
+  def is_exempt?(product)
+    products = %w(pills book books chocolate chocolates)
```
Comment (kalpana19sharma) [inline tax_determination.rb]: Declare this as a constant because this is not going to change.

File: tax_calculate.rb [hunk 27]
```diff
@@ -0,0 +1,25 @@
+class CalculateTax
+  WITH_TAX = []
+  @@sum_without_tax = 0.0
+  @@sum_with_tax = 0.0
+  @@diff = 0.0
+  def calculate_tax(last, tax)
+    @@sum_without_tax = @@sum_without_tax.to_f + last
+    @value = last + last * tax
+    @@sum_with_tax = @@sum_with_tax + @value 
+    CalculateTax::WITH_TAX << @value.round(2)
```
Comment (kalpana19sharma) [inline tax_calculate.rb]: We generally don't modify the value of constants.

File: new_arr.rb [hunk 28]
```diff
@@ -0,0 +1,20 @@
...
+  def return_result(arr)
+    arr.each do |i|
+      temp = ""
+      x = i.split(/ /)
+      for j in (0...x.length-2)
+        temp = temp + x[j] + " "
+      end
+      temp = temp + ": " + CreateArray::TAXES[@@j].to_s
+      @@j += 1
+      CreateArray::NEW_ARR << temp
```
Comment (kalpana19sharma) [inline new_arr.rb]: Don't modify the constants.

File: new_arr.rb [hunk 29]
```diff
@@ -0,0 +1,20 @@
...
+    end
+    CreateArray::NEW_ARR
+  end
+end
```
Comment (kalpana19sharma) [inline new_arr.rb]: Extra white spaces and no new line at the end of the file.

File: new_arr.rb [hunk 30]
```diff
@@ -0,0 +1,20 @@
+require_relative 'tax_calculate'
+module CreateArray
+  obj = CalculateTax.new
+  TAXES = obj.return_all_tax_value
+  NEW_ARR = []
+  @@j = 0
+  def return_result(arr)
+    arr.each do |i|
+      temp = ""
+      x = i.split(/ /)
```
Comment (kalpana19sharma) [inline new_arr.rb]: Use meaning full variable names.

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

**Review by kalpana19sharma — APPROVED** [review]

**Pull request merged**